### PR TITLE
feat(report): rework report generation flow

### PR DESCRIPTION
**Pull request by dpaolicelli-ktech**
Report generation is now a somewhat functional-style pipeline: time entry collection, data aggregation into a `tablib` `Databook` or `Dataset`, (optional) rendering.

As a proof-of-concept, this PR contains a reimplementation of the availability report.

### CODE REVIEW

File: src/krm3/core/models/auth.py [hunk 1]
```diff
@@ -209,16 +213,24 @@ def get_krm_days_with_contract(self, start_day: date, end_day: date) -> list[Krm
...
 
-    def get_contracts(self, start_day: date, end_day: date) -> list[Contract]:
+    def get_contracts(self, start_day: date, end_day: date | None, including_end: bool = True) -> list[Contract]:
         """Return a list of contracts applicable to the time interval between start_day and end_day."""
         from krm3.core.models import Contract  # noqa: PLC0415
 
         return list(
             Contract.objects.filter(
-                period__overlap=(start_day, end_day + datetime.timedelta(days=1) if end_day else None), resource=self
+                period__overlap=(start_day, end_day + datetime.timedelta(days=int(including_end)) if end_day else None),
```
Comment (dpaolicelli-ktech) [inline src/krm3/core/models/auth.py]: Note to self: we have `ContractQuerySet.active_between()`

File: src/krm3/reports/availability.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+class AbsenceKind(Enum):
+    HOLIDAY = gettext_lazy('H')
+    SICK = gettext_lazy('S')
+    LEAVE = gettext_lazy('L')
+    SPECIAL_LEAVE = gettext_lazy('SL')
+    REST = gettext_lazy('R')
+
+
+_ABSENCE_SHOW_HOURS = {AbsenceKind.LEAVE, AbsenceKind.SPECIAL_LEAVE, AbsenceKind.REST}
```
Comment (dpaolicelli-ktech) [inline src/krm3/reports/availability.py:27]: Note to self: what would be a better name?

File: tests/unit/reports/test_availability.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            project=None,
+        ).processed_data
+
+        row = dataset[0]
+        cell = row[1]
+        assert cell.date == datetime.date(2025, 8, 1)
+        assert cell.resource == resource
+        assert cell.absences == [Absence(AbsenceKind.HOLIDAY, Decimal(8))]
+
+    def test_resource_with_no_entries_has_empty_lists(self):
```
Comment (dpaolicelli-ktech) [inline tests/unit/reports/test_availability.py]: Note to self: maybe have the slop machine revise test names...

File: src/krm3/reports/availability.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+class Absence(NamedTuple):
+    kind: AbsenceKind
+    hours: Decimal
+
+    def __str__(self) -> str:
+        displayed_hours = self.hours if self.kind in _ABSENCE_SHOW_HOURS else ''
+        return f'{self.kind.value} {displayed_hours}'.rstrip()
+
+
+@dataclass
```
Comment (dpaolicelli-ktech) [inline src/krm3/reports/availability.py]: Note to self: maybe it should be `frozen = True`

~such immutability
much safety
wow~

File: src/krm3/reports/generator.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    def render[RD, F](self, renderer: Renderer[RD, F]) -> F:
+        """Transform `self.report_data` into a human-readable format.
+
+        :param renderer: The transformation `Callable` to apply
+        :return: The result of the transformation
+        """
+        return renderer(self.processed_data)
+
+    @property
+    def start(self) -> datetime.date:
```
Comment (dpaolicelli-ktech) [inline src/krm3/reports/generator.py]: Note to self: maybe these should be documented...

File: src/krm3/core/models/auth.py [hunk 6]
```diff
@@ -209,16 +212,24 @@ def get_krm_days_with_contract(self, start_day: date, end_day: date) -> list[Krm
 
         return days_list
 
-    def get_contracts(self, start_day: date, end_day: date) -> list[Contract]:
+    def get_contracts(self, start_day: date, end_day: date | None, including_end: bool = True) -> list[Contract]:
```
Comment (GabInTheCode) [inline src/krm3/core/models/auth.py:215]: Add context as to why you added including_end and why we are returning a list.

File: src/krm3/core/models/contracts.py [hunk 7]
```diff
@@ -21,16 +22,26 @@
 
 
 class ContractQuerySet(models.QuerySet['Contract']):
-    def active_between(self, start: datetime.date, end: datetime.date) -> Self:
+    def active_between(self, start: datetime.date, end: datetime.date | None, including_end: bool = True) -> Self:
```
Comment (GabInTheCode) [inline src/krm3/core/models/contracts.py:25]: Add context as to why you added including_end.

File: src/krm3/core/models/auth.py [hunk 8]
```diff
@@ -209,16 +212,24 @@ def get_krm_days_with_contract(self, start_day: date, end_day: date) -> list[Krm
...
         return list(
-            Contract.objects.filter(
-                period__overlap=(start_day, end_day + datetime.timedelta(days=1) if end_day else None), resource=self
-            )
+            cast('ContractQuerySet', Contract.objects)
+            .active_between(start_day, end_day, including_end)
+            .filter(resource=self)
         )
 
+    def get_contract_for_date(self, day: date | KrmDay) -> Contract | None:
```
Comment (GabInTheCode) [inline src/krm3/core/models/auth.py:238]: Document why it's the same day twice being passed.

File: src/krm3/reports/generator.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    @abc.abstractmethod
+    def collect(self, *args: P.args, **kwargs: P.kwargs) -> TimeEntryQuerySet:
+        """Collect the time entries on which the report should be generated.
+
+        :return: a Django queryset of `TimeEntry`.
+        """
+        ...
+
+    @abc.abstractmethod
+    def process(self, time_entries: TimeEntryQuerySet) -> RD:
```
Comment (GabInTheCode) [inline src/krm3/reports/generator.py]: Should probably be private methods

File: src/krm3/reports/generator.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        :return: a `tablib` databook or dataset with the processed data.
+        """
+        ...
+
+    def render[F](self, renderer: Renderer[RD, F]) -> F:
+        """Transform `self.report_data` into a different format.
+
+        :param renderer: The transformation `Callable` to apply
+        :return: The result of the transformation
+        """
```
Comment (GabInTheCode) [inline src/krm3/reports/generator.py]: docstring needs update

File: src/krm3/reports/generator.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        return renderer(self.processed_data)
+
+    @property
+    def start(self) -> datetime.date:
```
Comment (GabInTheCode) [inline src/krm3/reports/generator.py]: period start

File: src/krm3/reports/generator.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        """The start of the reporting period.
+
+        :return: a `datetime.date`.
+        """
+        return self.period[0]
+
+    @property
+    def end(self) -> datetime.date:
```
Comment (GabInTheCode) [inline src/krm3/reports/generator.py]: period end

File: src/krm3/utils/dates.py [hunk 13]
```diff
@@ -66,6 +65,9 @@ def is_holiday(self, country_calendar_code: str = None, include_sundays_as_holid
             return not get_country_holidays(country_calendar_code=country_calendar_code).is_working_day(self.date)
         return self.date in get_country_holidays(country_calendar_code=country_calendar_code)
 
+    def is_working_day(self, country_calendar_code: str | None = None) -> bool:
```
Comment (GabInTheCode) [inline src/krm3/utils/dates.py:68]: Document explaining why this will be important for the future change in the time entry

**Review by GabInTheCode — CHANGES_REQUESTED** [review]
Please complete the comments, and I will re-review.

File: tests/unit/reports/test_availability.py (hunk 3 above)
Comment (dpaolicelli-ktech) [reply tests/unit/reports/test_availability.py]: I just found out that this PR is missing tests with non-`None` projects. Thank you clanker...

**Comment by dpaolicelli-ktech** [comment]
Closing this - work has been moved to a different branch and ported to the new time entry models.
